crates/sui-types/src/waypoint.rs

```diff
@@ -0,0 +1,426 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{Deserialize, Serialize};
+use std::borrow::Borrow;
+use std::fmt::Debug;
+use thiserror::Error;
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use curve25519_dalek::ristretto::RistrettoPoint;
+use ed25519_dalek::Sha512;
+
+use crate::{base_types::AuthorityName, committee::Committee};
+
+#[cfg(test)]
+#[path = "unit_tests/waypoint_tests.rs"]
+mod waypoint_tests;
+
+#[derive(Eq, PartialEq, Clone, Debug, Error)]
+#[allow(clippy::large_enum_variant)]
+pub enum WaypointError {
+    #[error("Waypoint error: {:?}", msg)]
+    Generic { msg: String },
+    #[error("Nothing to do, all parts are already in the checkpoint.")]
+    NothingToDo,
+    #[error("Fragment cannot connect with global checkpoint.")]
+    CannotConnect,
+}
+
+impl WaypointError {
+    pub fn generic(msg: String) -> WaypointError {
+        WaypointError::Generic { msg }
+    }
+}
+
+/*
+   A MulHash accumulator: each element is mapped to a
+   point on an eliptic curve on which the DL problem is
+   hard. The accumulator is the sum of all points.
+*/
+#[derive(Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Accumulator {
+    accumulator: RistrettoPoint,
+}
+
+impl Accumulator {
+    /// Insert one item in the accumulator
+    pub fn insert<I>(&mut self, item: &I)
+    where
+        I: Borrow<[u8]>,
+    {
+        let point = RistrettoPoint::hash_from_bytes::<Sha512>(item.borrow());
+        self.accumulator += point;
+    }
+
+    // Insert all items from an iterator into the accumulator
+    pub fn insert_all<'a, I, It>(&'a mut self, items: It)
```

Comment:
I believe you're looking for an implementation of [`Sum`](https://doc.rust-lang.org/std/iter/trait.Sum.html), which is why I was suggesting `std::ops::Add`  earlier.

Reply:
Yep -- still lets keep this hand written for the moment, since I really want to stay away from making this look like a type anyone should use for anything else.